src/modules/samples/cpp/src/so/SampleModule.cpp

```diff
@@ -0,0 +1,395 @@
+{
+    OsConfigLogInfo(SampleLog::Get(), "C++ Sample module unloaded");
+    SampleLog::CloseLog();
+}
+
+int MmiGetInfo(
+    const char* clientName,
+    MMI_JSON_STRING* payload,
+    int* payloadSizeBytes)
+{
+    int status = MMI_OK;
+
+    constexpr const char ret[] = R""""({
+        "Name": "C++ Sample",
+        "Description": "A sample module written in C++",
+        "Manufacturer": "Microsoft",
+        "VersionMajor": 1,
+        "VersionMinor": 0,
+        "VersionInfo": "",
+        "Components": ["SampleComponent"],
+        "Lifetime": 1,
+        "UserAccount": 0})"""";
+
+    ScopeGuard sg{[&]()
+    {
+        if (MMI_OK == status)
+        {
+            if (IsFullLoggingEnabled())
+            {
+                OsConfigLogInfo(SampleLog::Get(), "MmiGetInfo(%s, %.*s, %d) returned %d", clientName, *payloadSizeBytes, *payload, *payloadSizeBytes, status);
+            }
+            else
+            {
+                OsConfigLogInfo(SampleLog::Get(), "MmiGetInfo(%s, -, %d) returned %d", clientName, *payloadSizeBytes, status);
+            }
+        }
+        else
+        {
+            if (IsFullLoggingEnabled())
+            {
+                OsConfigLogError(SampleLog::Get(), "MmiGetInfo(%s, %.*s, %d) returned %d", clientName, *payloadSizeBytes, *payload, *payloadSizeBytes, status);
+            }
+            else
+            {
+                OsConfigLogError(SampleLog::Get(), "MmiGetInfo(%s, -, %d) returned %d", clientName, *payloadSizeBytes, status);
+            }
+        }
+    }};
+
+    if (nullptr == clientName)
+    {
+        OsConfigLogError(SampleLog::Get(), "MmiGetInfo called with null clientName");
+        status = EINVAL;
+    }
+    else if (nullptr == payload)
+    {
+        OsConfigLogError(SampleLog::Get(), "MmiGetInfo called with null payload");
+        status = EINVAL;
+    }
+    else if (nullptr == payloadSizeBytes)
+    {
+        OsConfigLogError(SampleLog::Get(), "MmiGetInfo called with null payloadSizeBytes");
+        status = EINVAL;
+    }
+    else
+    {
+        try
+        {
+            std::size_t len = ARRAY_SIZE(ret) - 1;
+            *payload = new (std::nothrow) char[len];
+            if (nullptr == *payload)
+            {
+                OsConfigLogError(SampleLog::Get(), "MmiGetInfo failed to allocate memory");
+                status = ENOMEM;
+            }
+            else
+            {
+                std::memcpy(*payload, ret, len);
+                *payloadSizeBytes = len;
+            }
+        }
+        catch (const std::exception& e)
+        {
+            OsConfigLogError(SampleLog::Get(), "MmiGetInfo exception thrown: %s", e.what());
+            status = EINTR;
+
+            if (nullptr != *payload)
+            {
+                delete[] * payload;
+                *payload = nullptr;
+            }
+
+            if (nullptr != payloadSizeBytes)
+            {
+                *payloadSizeBytes = 0;
+            }
+        }
```

Comment:
This should be into the base class GetInfo member. MmiGetInfo should just invoke that. See modules.md